### PR TITLE
clean up golang memory leaks

### LINKED ISSUE #944: Memory Leak in Golang
Hi OsoHQ,

Been enjoying the library, and I hope to use it for an authorization service, however I believe I found a memory leak in the Golang library.

## Setup

I was able to reproduce using the `oso-go-quickstart` example. I [forked that repo](https://github.com/bhb603/oso-go-quickstart) and the only changes I made were:
- upgraded to the latest oso  `v0.11.3 ->  v0.12.3` (https://github.com/bhb603/oso-go-quickstart/blob/main/go.mod#L5)
- added a policy:
    ```
    allow(actor: String, "GET", expense: Expense) if  
            expense.SubmittedBy = actor;
    ```
- [Dockerized](https://github.com/bhb603/oso-go-quickstart/blob/main/Dockerfile) the app and deployed it to a kubernetes cluster with [this simple deployment](https://github.com/bhb603/oso-go-quickstart/blob/main/k8s/deployment.yaml) (my ultimate use case would be to deploy in kubernetes, which is why I replicated that here)

## Testing

I have some load-testing tooling which can generate requests against the service. Since this quickstart example only has one endpoint, it just requests:
```
curl -H "user: alice@example.com" http://oso-go-quickstart.oso:5050/expenses/1
```
Which would check `oso.IsAllowed(User("alice@example.com"), "GET", expense)` and return:
```
HTTP/1.1 200 OK
Content-Length: 40
Content-Type: text/plain; charset=utf-8

Expense{{500 coffee alice@example.com}}
```

## Results

Memory Usage increases linearly when the service is under load, until it reaches the limit I set for the container (256M) at which point Kubernetes kills the container (OOMKIlled) and it restarts, producing the saw tooth graph below.

The rate of memory increase is proportional to the load (reqs per second), i.e. it increases ~2.5x faster when I was hitting it at 25rps vs 10rps, and flattens out with 0rps.

Memory usage under load:
Containers___Datadog

More Metrics:
image

### PR DESCRIPTION
**Pull request by saolsen**
Clean up golang memory leaks.
Wasn't calling query_free in all cases, wasn't calling string_free in some cases and wasn't freeing Cstrings allocated from go to pass through ffi calls.

fixes #944

### CODE REVIEW

File: languages/go/internal/ffi/ffi.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ func (p *PolarFfi) delete() {
 func getError() error {
 	err := C.polar_get_error()
 	errStr := C.GoString(err)
```
Comment (samscott89) [inline languages/go/internal/ffi/ffi.go:47]: Minor: could move all go string construction into a method so we don't forget it in the future.

File: languages/go/internal/ffi/ffi.go [hunk 2]
```diff
@@ -131,6 +137,8 @@ func (p PolarFfi) NewQueryFromTerm(queryTerm types.Term) (*QueryFfi, error) {
 	json, err := ffiSerialize(queryTerm)
 	if err != nil {
 		return nil, err
+	} else {
+		defer C.free(unsafe.Pointer(json))
```
Comment (samscott89) [inline languages/go/internal/ffi/ffi.go:141]: It's actually fine to call free on a nil pointer. So I would prefer we keep all the frees right next to where we get the string back. Will be easier to scan

Comment (samscott89) [reply languages/go/internal/ffi/ffi.go:141]: And avoid the else blocks that are a _little_ hard to follow.

File: languages/go/internal/ffi/ffi.go [hunk 3]
```diff
@@ -152,10 +160,12 @@ func (p PolarFfi) NextInlineQuery() (*QueryFfi, error) {
 
 func (p PolarFfi) RegisterConstant(term types.Term, name string) error {
 	cName := C.CString(name)
+	defer C.free(unsafe.Pointer(cName))
 	cTerm, err := ffiSerialize(term)
 	if err != nil {
-		defer C.free(unsafe.Pointer(cName))
 		return err
+	} else {
```
Comment (samscott89) [inline languages/go/internal/ffi/ffi.go:167]: ditto

File: languages/go/internal/ffi/ffi.go [hunk 4]
```diff
@@ -191,6 +201,8 @@ func (q QueryFfi) CallResult(callID uint64, term *types.Term) error {
 		s, err = ffiSerialize(term)
 		if err != nil {
 			return err
+		} else {
```
Comment (samscott89) [inline languages/go/internal/ffi/ffi.go:204]: ditto

File: languages/go/query.go [hunk 5]
```diff
@@ -128,6 +132,7 @@ func (q *Query) Next() (*map[string]interface{}, error) {
 			return nil, fmt.Errorf("unexpected query event: %v", ev)
```
Comment (samscott89) [inline languages/go/query.go:132]: Cleanup here too? Although memory leak wouldnt be the worst issue if this was hit a lot...

File: languages/go/oso.go [hunk 6]
```diff
@@ -142,6 +142,7 @@ func (o Oso) IsAllowed(actor interface{}, action interface{}, resource interface
 	if err != nil {
 		return false, err
 	} else if results != nil {
+		defer query.Cleanup()
```
Comment (samscott89) [inline languages/go/oso.go:145]: It's very non-obvious why this is needed. Can we add a comment here and to the `Next` method stating that if you don't exhaust the iterator, you need to call `Cleanup` manually?

**Review by samscott89 — CHANGES_REQUESTED** [review]
Looks good. I have a few suggestions to help avoid this in the future.

File: languages/go/query.go (hunk 5 above)
Comment (saolsen) [reply languages/go/query.go:132]: Yeah this is really like a panic case. Shouldn't ever happen.

**Review by samscott89 — APPROVED** [review]
